src/build_log.cc

```diff
@@ -377,7 +397,11 @@ LoadStatus BuildLog::Load(const string& path, string* err) {
 BuildLog::LogEntry* BuildLog::LookupByOutput(const string& path) {
   Entries::iterator i = entries_.find(path);
   if (i != entries_.end())
+#if __cplusplus >= 201103L || (defined(_MSC_VER) && _MSC_VER >= 1900)
+    return i->second.get();
```

Comment:
i think that for many of the places where you are returning `ptr.get()` you would instead want to change the function to return `LogEntryPtr const&`

Reply:
Hey @jonesmz,

This function also returns an rvalue, so your suggestion will cause a segmentation fault:

https://godbolt.org/z/a99acs3oe

Reply:
That's because functions which return a const& may not return a variable that lives only in the stack of that function.

My recommendation is that for functions which are returning items that live inside of a map, you can return the mapped item as const&, because the map itself won't be deleteing the item until something tells it so.

But you can't change the return type to std::unique_ptr const&, and then return a c-style pointer, because that will result in returning a temporary variable that lives on the stack of the function as a const&, aka a dangling reference.

I wasn't specifically telling you that you *must* do any of the things I commented about. I'm just giving high level design feedback. Don't take my comments as word-of-law, and don't assume that I spent a whole lot of time looking through the code to make sure my advice was 100% correct. It was a drive by recommendation based on years of working with the smart pointers from the standard library.

Reply:
Or much better way would be using `std::optional`, but it is implemented since C++17...